flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/NumericToTimestampCastRule.java

```diff
@@ -0,0 +1,67 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.planner.functions.casting;
+
+import org.apache.flink.table.api.ValidationException;
+import org.apache.flink.table.data.TimestampData;
+import org.apache.flink.table.types.logical.LogicalType;
+import org.apache.flink.table.types.logical.LogicalTypeFamily;
+import org.apache.flink.table.types.logical.LogicalTypeRoot;
+
+/**
+ * {@link LogicalTypeFamily#NUMERIC} to {@link LogicalTypeRoot#TIMESTAMP_WITHOUT_TIME_ZONE}/{@link
+ * LogicalTypeRoot#TIMESTAMP_WITH_LOCAL_TIME_ZONE} cast rule. Disable cast conversion between
+ * Numeric type and Timestamp type and suggest to use {@code TO_TIMESTAMP()}/{@code
+ * TO_TIMESTAMP_LTZ()} instead.
+ */
+class NumericToTimestampCastRule
+        extends AbstractExpressionCodeGeneratorCastRule<Number, TimestampData> {
+
+    static final NumericToTimestampCastRule INSTANCE = new NumericToTimestampCastRule();
+
+    private NumericToTimestampCastRule() {
+        super(
+                CastRulePredicate.builder()
+                        .input(LogicalTypeFamily.NUMERIC)
+                        .target(LogicalTypeRoot.TIMESTAMP_WITHOUT_TIME_ZONE)
+                        .target(LogicalTypeRoot.TIMESTAMP_WITH_LOCAL_TIME_ZONE)
+                        .build());
+    }
+
+    @Override
+    public String generateExpression(
+            CodeGeneratorCastRule.Context context,
+            String inputTerm,
+            LogicalType inputLogicalType,
+            LogicalType targetLogicalType) {
+        if (targetLogicalType.is(LogicalTypeRoot.TIMESTAMP_WITHOUT_TIME_ZONE)) {
+            throw new ValidationException(
+                    "The cast conversion from NUMERIC type to TIMESTAMP type"
+                            + " is not allowed, it's recommended to use TO_TIMESTAMP(FROM_UNIXTIME(numeric_col))"
+                            + " instead, note the numeric is in seconds.");
+        } else if (targetLogicalType.is(LogicalTypeRoot.TIMESTAMP_WITH_LOCAL_TIME_ZONE)) {
+            throw new ValidationException(
+                    "The cast conversion from NUMERIC type"
+                            + " to TIMESTAMP_LTZ type is not allowed, it's recommended to use"
+                            + " TO_TIMESTAMP_LTZ(numeric_col, precision) instead.");
+        } else {
+            throw new IllegalArgumentException("This is a bug. Please file an issue.");
+        }
```

Comment:
Given we're touching this, can we move this to somewhere like `LogicalTypeCasts` or similar where it better belongs to? Given we're also ending up linking `LogicalTypeCasts` with calcite (https://github.com/apache/flink/pull/18524/files#diff-8c1c5cdfd4ee1de9d54fd39db308ab1d7fa9fddb2db1e36cdd1c9b0b8fc90f4bR173), we could have a method there like `getFailureMessage(LogicalType in, LogicalType out)` or similar which contains all the logic for the special "error messages" for cast tuples not matching.

Also the problem with failing in the rule is that we assume all around the code that `generate` never fails if `exists()` returns true for that rule, so i'm not sure if this is safe to do.

Reply:
Please check: https://github.com/apache/flink/pull/18582/commits/29d8d03b1be818a64834e5ba670a83d8857111ab
If it's too hacky, maybe we can postpone this refactoring after https://github.com/apache/flink/pull/18524, WDYT @slinkydeveloper ?

Reply:
let's wait for #18524 to get in first

Reply:
Ok, I will remove that commit for the moment

Reply:
Opened issue: https://issues.apache.org/jira/browse/FLINK-25928

Reply:
So now your strategy is to merge it this way and then we fix it with FLINK-25928, right?

Reply:
yep!